man/njeroutes.8
.\" $Header$
.\"
.\"  Man page for HUJI-NJE/FUNET-NJE on UNIX system.
.\"
.\"  Written by  Matti Aarnio <mea@finfiles.bitnet> <mea@nic.funet.fi>
.\"
.\"  Date: 27-Dec-1990, 18-Sep-1993, 22-Oct-1993
.\"
.fp 4 CW
.ds ]W Funet-Nje 8 UNIX 3.0
.TH NJEROUTES 8
.SH NAME
njeroutes \- FUNETNJE bitnet software routing database generator.
.SH SYNOPSIS
njeroutes
.I header.filename
.I routes.filename
.I database.filename
.SH DESCRIPTION
.IX "bitnet"
.IX "funetnje"
.IX "njeroutes"
.LP
.I Net_routes
builds routing database of BITNET nodes for
.I funetnje
program internal usage.
If no
\f4DEFAULT-ROUTE
has been defined in
.I /etc/funetnje.cf
(see
.BR funetnje (8))
then this is ONLY way to route messages and files from local node
to anywhere.
.LP
Example of
.I header.file
for node \f4FINFILES\fR having alias name \f4HAMSTER\fR.
.LP
.nf
\f4*
\f4*  Header of HUyNJE/FUNETNJE package routing table
\f4*
\f4*  Generic format:
\f4*	SITE	LINE	(FORMAT/INFO)
\f4*
\f4 ROUTE FINFILES LOCAL ASCII
\f4 ROUTE HAMSTER  LOCAL ASCII
.fi
.LP
It may also be useful to define alternate routes just in
case the direct link doesn't work:
.LP
.nf
\f4 ROUTE FIGBOX FIGBOX
\f4 ROUTE FINHUTC FIGBOX
.fi
.LP
First one defines that \f4FIGBOX\f4 has no alternatives, while
\f4FINHUTC\fR can get routed via \f4FIGBOX\fR in case direct
link to it is not active.
.LP
Example routing data for same node.  Used format is same as
IBM RSCS v1 uses, thus it is easily obtainable using common BITNET
tools.
Next step connection of the example node is \f4FIGBOX\fR.
.LP
.nf
\f4* * The position of the network name and country *
\f4* * assumes that the RSCS ROUTE command does not *
\f4* * have more than two parameters.               *
\f4ROUTE VERS9012 FIGBOX   EARN     NL
\f4ROUTE LINKS012 FIGBOX   BITNET   US
\f4ROUTE FINABO   FIGBOX   EARN     FI
\f4ROUTE FINALKO  FIGBOX   EARN     FI
\f4ROUTE FINFUN   FIGBOX   EARN     FI
\f4ROUTE FINGATE  FIGBOX   EARN     FI
\f4ROUTE FINHUT   FIGBOX   EARN     FI
.fi
.LP
.SH SEE ALSO
.BR funetnje (8),
.BR send (1),
.BR sendfile (1),
.BR bmail (8),
.BR ucp (8).
.SH BUGS
.LP
None, we hope.
.SH DIAGNOSTICS
All error conditions are reported with self explanatory text messages.
.LP
If multiple entries of same node are attempted (for example one in
header, other in routes), first one stays as routing definition, rest
are reported as duplicates.
.SH WARNING
It is not advised to create output on same
.BR file ,
as the current database is, although they will not be intermixed, as
if there is a file with the target name, it will be unlinked before
the processing starts.
.LP
Once the database is generated, and placed (with
.BR mv (1),
for example) on place of the active database, a
.RB ` ucp "(8) " "rescan route" '
-command is suggested (or full restart of the emulator.)
.SH AUTHORS
.LP
Yehavi Bourvine, Hebrew University of Yerusalem, Israel
.I <yehavi@hujivms.bitnet>.
.LP
Matti Aarnio, Finnish University and Research Network \- FUNET, Finland
.I <mea@nic.funet.fi> <mea@finfiles.bitnet>.
.LP
Original (VMS oriented package) software by Yehavi Bourvine,
UNIX features, plenty of modifications, man-pages by Matti Aarnio.


ChangeLog
Fri Mar 22 13:48:30 1996  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    Even more relaxed approach on parametrizing
	    the CMDMAILBOX -- now allows SOCK/FIFO/UDP
	    with fairly flexible (in regards of spaces)
	    syntax..

Mon Nov 13 17:48:52 1995  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    A bit more relaxed approach on parametrizing
	    CMDMAILBOX config parameter (FIFO/SOCKET case)

	* file_queue.c:
	    Do RSCS-accounting on locally sent files, that
	    wind up to local recipients

	* unix_brcst.c:
	    Re-enter the SysVR4 UTMP weirdos..  Now fixed ?

Fri Aug 25 09:02:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* mailify.c:
	    Handling error-messages (where "MAIL FROM:" address
	    is "<>", that is, empty) was done improperly.  Now
	    it does flag input as from "channel error" in that
	    case.

Mon Jul 24 14:29:59 1995  Matti Aarnio  <mea@nic.funet.fi>

	* bmail.c, sendfile.c:
	    Sometimes   mktemp()  fails when there are a LOT of
	    files being sent at the same time.  Changed now to
	    more classical  ".MaiL.%05d"-format with  getpid()
	    result as integer data..

Wed Jun 21 22:42:39 1995  Matti Aarnio  <mea@nic.funet.fi>

	* unix_brcst.c, nmr_unix.c:
	    Ok, more than __linux__ is needed to know which
	    machines have <utmp.h> containing extended flags.
	    Now we test  "DEAD_PROCESS" pre-processor define,
	    which existence (propably) will correlate with
	    the existence of utmp's field  ut_type.
	    (SysVr4 feature, I think.)

Mon Jun 19 15:17:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* send.c, sendfile.c, bmail.c:
	    Some additional tests to trap various forms of bad
	    addresses -- accepts only  user@node format with
	    no '.' in the address, nor either component longer
	    than 8 chars.

	* nmr.c:
	    Made sure not to overwrite allocated string spaces on
	    Msg-queue entry.  (Did cause a crash with it..)

Tue Jun 13 09:13:18 1995  Matti Aarnio  <mea@nic.funet.fi>

	* qrdr.c:
	    Collect directory browsing results to an array at first,
	    then qsort() it with the time-stamp, and finally display
	    them in order. (Latest last)
	    (This is something what Gerald Hanush asked for.)

Mon Jun 12 22:44:00 1995  Matti Aarnio  <mea@nic.funet.fi>

	* file_queue.c:
	    Apparently OSF/1 pipe() return value is the first
	    fd, not "0", when it succeeds.  Nice thing - and not..
	    Altered the test to see if it returns a value under
	    zero, or something positive.  When the return is
	    positive (zero, or larger), fork() and run in
	    child-mode.

	    If a child_pipe exists already ( >= 0 ), close
	    it and mark the new one for processing.

	* libreadcfg.c, read_config, libsendcmd.c, unix.c:
	    Rewrote things so that now the used CMDMAILBOX
	    transport method is chosen at runtime:

	    CMDMAILBOX  F  /path/to/.cmdfifo
	    CMDMAILBOX  f  /path/to/.cmdfifo
	    CMDMAILBOX  S  /path/to/.cmdsocket
	    CMDMAILBOX  U  127.0.0.2 175	# UDP port

	    The "f" is FIFO which is closed, and reopened
	    every time one command is pulled in.

Sun Jun 11 03:12:06 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesparser.c:
	    Noticed that the data parsing stopped at first ":",
	    which caused premature breaks; observe following entry
	    from BITEARN.NODES -file:
	    	:remark1.BASE(9501) 00/00/00 00:00:00

	* libreadcfg.c, read_config.c, libsendcmd.c, unix.c:
	   Amended "CMDMAILBOX"-data to contain port-number in
	   case UDP-format is used. (That is, have two tokens.)
	   Default value for the port-number is 175.

	* unix_tcp.c read_passive_tcp_connection():
	   Noticed that a case when a socket exists, and a new
	   connection is attempted is handled improperly. The
	   old socket-fd was FORGOTTEN!  Now it NAKs the connection
	   attempt with a complaint that there already is a connection
	   in there...

	   Things will progress from such a situation usually by
	   detecting a closure on the existing stream, but better
	   not to rush ahead of things.

	* bmail.c:
	   Accept "-u" only when calling the program from uids 0 or 1
	   (root and daemon in most common cases -- for BSD-sendmail
	    which wants  bmail executable with protection 2111 (user X,
	    group NJEGRP) )

Sat Jun 10 05:18:27 1995  Matti Aarnio  <mea@nic.funet.fi>

	* Makefile, consts.h, prototypes.h, recv_file.c, io.c,
	  send_file.c, unix.c, protocol.c, unix_tcp.c:

	    - Found one uninitialized pointer, which apparently
	      caused mystic data-structure corruptions (on timer
	      subsystem)

	    - Made mandatory (well, with disabling options) to
	      use NBSTREAM, USE_XMIT_QUEUE, and USE_SOCKOPT.
	      (in consts.h, see also  Makefile's  comments!)

	    - Use  F_WAIT_V_A_BIT  to flag if there is any reason
	      for NOT to send out data at the moment, and instead
	      quit from within  handle_ack()..
	      The F_WAIT_V_A_BIT is set when new data is sent to
	      outbound buffer (primary one, or XMIT QUEUE).
	      The bit is cleared when pending writer finishes
	      with success status.  When the pending-writer is
	      called from  poll_sockets(), immediately after it
	      the  dequeue_xmit_queue() is called, and if there
	      is something, it can cause F_WAIT_V_A_BIT to be
	      set again..)

	    - If the XMIT_QUEUE ever runs out of space, it abort()s!

	    - Tested the beast by sending 80 or so smallish files both
	      ways over the link.  Such setup caused previously link
	      corruption..  Actually the critical phase is when there
	      are flowing NJE headers/trailer, bare data sending works
	      just fine, but when a file ends, EOF block needs an ack
	      to be sent, and unless there the care is used, all
	      queues will overflow..

Thu Jun  8 15:48:14 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesfilter.c:
	    New option: "-netsoft" to list node software fields.

	* funetnje.texinfo:
	    Some additional chapters

	* unix_tcp.c:

njeroutes.c
/* NJEROUTES.C 1.3-mea
 |
 |  Build up the route database for FUNET-NJE.
 |  Copyright (C) Finnish University and Research Network, FUNET
 |  1991,1993,1994
 |
 |  There is marginal resemblance with original UNIX_BUILD of
 |  HUJI-NJE, however this is no longer such a a primitive thing,
 |  as that was...
*/

#include "consts.h"
#include "prototypes.h"
#include "bintree.h"

/* extern char *sys_errlist[]; */
extern int errno;

char *pname;

int CompareRoute(r1,r2)
struct ROUTE_DATA *r1, *r2;
{
	return strcmp(r1->DestNode, r2->DestNode);
}

int
main(argc,argv)
     int argc;
     char *argv[];
{
	char	*c, line[132],
		site[21], LineName[21], format[20],
		temp_line[132];	/* Just a place holder */
	char	in_file[80], header_file[80], out_file[80];
	FILE	*Ifd, *fd, *Dup=NULL;
	int	counter, Dcount, lines;
	struct ROUTE_DATA  RouteRecord;
	struct Bintree *BT;

	if ((pname = strrchr(argv[0],'/')) != NULL)
	  ++pname;
	else
	  pname = argv[0];

/* Get files' names */
	if (argc != 4) {
	  printf("%s  Creates FUNET-NJE routing tables.\n",pname);
	  printf("%s  Given 3 arguments the interactive mode is not activated:\n",pname);
	  printf("%s  Args are in order:  Header-Fname, RoutingTbl-Fname, OutputFname\n",pname);
	  

	  printf("Header file: "); gets(header_file);
	  printf("Routing table: "); gets(in_file);
	  printf("Output file: "); gets(out_file);
	} else  {
	  strncpy( header_file,argv[1],sizeof header_file -1);
	  strncpy( in_file,argv[2],sizeof in_file -1);
	  strncpy( out_file,argv[3],sizeof out_file -1);
	}	  
	
	if ((c = strchr(header_file, '\n')) != NULL) *c = '\0';
	if ((c = strchr(in_file, '\n')) != NULL) *c = '\0';
	if ((c = strchr(out_file, '\n')) != NULL) *c = '\0';

	if ((Ifd = fopen(header_file, "r")) == NULL) {
	  printf("Can't open header file '%s'\n", header_file);
	  exit(1);
	}
	if ((fd = fopen(in_file, "r")) == NULL) {
	  printf("Can't open input file: %s\n", in_file);
	  exit(1);
	}

	/* Open the database file */

	/* Make sure it doesn't exist now.. */
	unlink(out_file);
	BT = bintree_open(out_file,
			  sizeof(struct ROUTE_DATA),
			  CompareRoute);
	if (BT == NULL) {
	  int oerrno = errno;
	  fprintf(stderr,"%s:  Open of `%s' database failed because of error: %s\n",
		  pname,out_file,PRINT_ERRNO);
	  exit(oerrno);
	}

	counter = 0;  Dcount = 0;
	lines = 0;
	printf("Records read:\n");

	/* The header file */

	while (fgets(line, sizeof line, Ifd) != NULL) {
	  ++lines;
	  if ((c = strchr(line, '\n')) != 0) *c = 0;
	  /*printf("Got line: '%s'\n",line);*/
	  if (*line == '*') continue; /* Comment line */
	  if (*line == '#') continue; /* Comment line */
	  if (*line == 0 )  continue; /* Empty line */
	  counter++;
	  if (counter % 100 == 0) {
	    printf("%d%s", counter,(counter%1000 == 0) ? "\n":" ");
	    fflush(stdout);
	  }
	  if (sscanf(line, "%*s %s %s %s",
		     site, LineName, format) < 3) {
	    printf("Illegal line: %s\n", line);
	    continue;
	  }
	  if (strcasecmp(LineName,"DEFAULT")==0 ||
	      strcasecmp(LineName,"SYSTEM")==0  ||
	      strcasecmp(LineName,"OFF")== 0) {
	    printf("Bad route target: %d: '%s'\n",lines, line);
	    continue;
	  }

	  memcpy(RouteRecord.DestNode,site,8);
	  RouteRecord.DestNode[8] = 0;
	  memcpy(RouteRecord.DestLink,LineName,8);
	  RouteRecord.DestLink[8] = 0;
	  RouteRecord.Fmt = *format;

	  if (bintree_find(BT,&RouteRecord) == NULL) {
	    /* Not found already, insert! */
	    bintree_insert(BT,&RouteRecord);
	  } else {
	    /* A duplicate ! */
	    ++Dcount;
	    if (!Dup) Dup = fopen("DUPLICATES.TMP","w");
	    if (Dup) fprintf(Dup,"HdrDup: %d: %s\n",lines, line);
	  }
	}

	printf("Processing routing table\n");
	
	/* Now process the routing table */

	lines = 0;
	while (fgets(line, sizeof line, fd) != NULL) {
	  ++lines;
	  if ((c = strchr(line, '\n')) != 0)
	    *c = 0;
	  /*printf("Got line: '%s'\n",line);*/
	  if (*line == '*') continue; /* Comment line */
	  if (*line == '#') continue; /* Comment line */
	  if (*line == 0 ) continue; /* Empty line */
	  counter++;
	  if (counter % 100 == 0) {
	    printf("%d%s", counter,(counter%1000 == 0) ? "\n":" ");
	    fflush(stdout);
	  }
	  /* 'ROUTE node ToLink' */
	  if (sscanf(line, "%s %s %s",
		     temp_line, site, LineName) < 2) {
	    printf("Illegal line: %s\n", line);
	    continue;
	  }
	  if (strcasecmp(LineName,"DEFAULT")==0 ||
	      strcasecmp(LineName,"SYSTEM")==0  ||
	      strcasecmp(LineName,"OFF")==0) {
	    printf("Bad route target: '%s'\n",line);
	    continue;
	  }
	  memcpy(RouteRecord.DestNode,site,8);
	  RouteRecord.DestNode[8] = 0;
	  memcpy(RouteRecord.DestLink,LineName,8);
	  RouteRecord.DestLink[8] = 0;
	  RouteRecord.Fmt = *format;

	  if (bintree_find(BT,&RouteRecord) == NULL) {
	    /* Not found already, insert! */
	    bintree_insert(BT,&RouteRecord);
	  } else {
	    /* A duplicate ! */
	    ++Dcount;
	    if (!Dup) Dup = fopen("DUPLICATES.TMP","w");
	    if (Dup) fprintf(Dup,"RteDup: %d: %s\n",lines, line);
	  }
	}

	fclose(fd);
	fclose(Ifd);
	bintree_close(BT);

	printf("\nTotal records inserted: %d, %d duplicates in DUPLICATE.TMP\n",
	       counter, Dcount);
	if (Dup) fclose(Dup);
	return 0;
}

